builder/docker/docker.go

```diff
@@ -0,0 +1,350 @@
+package docker
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Masterminds/cookoo"
+	"github.com/Masterminds/cookoo/log"
+	"github.com/Masterminds/cookoo/safely"
+	"github.com/deis/deis/builder/etcd"
+	docli "github.com/fsouza/go-dockerclient"
+)
+
+// Path to the Docker unix socket.
+var DockSock = "/var/run/docker.sock"
+
+// Cleanup removes any existing Docker artifacts.
+//
+// Returns true if the file exists (and was deleted), or false if no file
+// was deleted.
+func Cleanup(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+
+	// If info is returned, then the file is there. If we get an error, we're
+	// pretty much not going to be able to remove the file (which probably
+	// doesn't exist).
+	if _, err := os.Stat(DockSock); err == nil {
+		log.Infof(c, "Removing leftover docker socket %s: %v", DockSock)
+		return true, os.Remove(DockSock)
+	}
+	return false, nil
+}
+
+// CreateClient creates a new Docker client.
+//
+// Params:
+// 	- url (string): The URI to the Docker daemon. This defaults to the UNIX
+// 		socket /var/run/docker.sock.
+//
+// Returns:
+// 	- *docker.Client
+//
+func CreateClient(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	path := p.Get("url", "unix:///var/run/docker.sock").(string)
+
+	return docli.NewClient(path)
+}
+
+// Start starts a Docker daemon.
+//
+// This assumes the presence of the docker client on the host. It does not use
+// the API.
+func Start(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+	dargs := []string{
+		"-d",
+		"--bip=172.19.42.1/16",
+		"--insecure-registry",
+		"10.0.0.0/8",
+		"--insecure-registry",
+		"172.16.0.0/12",
+		"--insecure-registry",
+		"192.168.0.0/16",
+		"--insecure-registry",
+		"100.64.0.0/10",
```

Comment:
This is basically starting the registry allowing it to accept connections from AWS network, bare metal network and other networks. That's fine but it means that we have to maintain it every time a new CIDR shows up. Not sure where the right place for these either. @carmstrong ?